core/arch/arm/mm/core_mmu_lpae.c

```diff
@@ -121,10 +121,22 @@
 #define LOWER_ATTRS(x)			(((x) & 0xfff) << LOWER_ATTRS_SHIFT)
 
 #define ATTR_DEVICE_INDEX		0x0
+#define ATTR_DEVICE_nGnRE_INDEX		0x0
 #define ATTR_IWBWA_OWBWA_NTR_INDEX	0x1
+#define ATTR_DEVICE_IO_INDEX		0x2
+#define ATTR_DEVICE_nGnRnE_INDEX	0x3
+#define ATTR_DEVICE_nGRE_INDEX		0x4
+#define ATTR_DEVICE_GRE_INDEX		0x5
 #define ATTR_INDEX_MASK			0x7
 
+#define ATTR_DEVICE_MASK		(0xf)
 #define ATTR_DEVICE			(0x4)
+
+#define ATTR_DEVICE_nGnRnE		(0x0)
```

Comment:
Do we need all these types? So far we've been able to keep the number of different types down by only adding what's really needed.

Reply:
The spec allows normally all of them. But I am checking internally if we need to support all of them.

Reply:
Thanks.

Reply:
I checked and we can give an error back for the none supported once. However the spec states:

Reply:
OK, no problem.

Reply:
I update the patch and only allow nGnRE and nGnRnE